src/backend/commands/dirtablecmds.c

```diff
@@ -0,0 +1,472 @@
+		AclResult	aclresult;
+
+		aclresult = pg_tablespace_aclcheck(tablespaceId, GetUserId(),
+									 		ACL_CREATE);
+		if (aclresult != ACLCHECK_OK)
+			aclcheck_error(aclresult, OBJECT_TABLESPACE,
+				  			get_tablespace_name(tablespaceId));
+	}
+
+	/* In all cases disallow placing user relations in pg_global */
+	if (tablespaceId == GLOBALTABLESPACE_OID)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("only shared relations can be placed in pg_global tablespace")));
+
+	return tablespaceId;
+}
+
+void
+CreateDirectoryTable(CreateDirectoryTableStmt *stmt, Oid relId)
+{
+	Relation 	dirRelation;
+	Datum 		values[Natts_pg_directory_table];
+	bool 		nulls[Natts_pg_directory_table];
+	HeapTuple	tuple;
+	char 		*dirTablePath;
+	Form_pg_class pg_class_tuple;
+	HeapTuple	class_tuple;
+	Oid 		spcId = chooseTableSpace(stmt);
+
+	class_tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(relId));
+	if (!HeapTupleIsValid(class_tuple))
+		elog(ERROR, "cache lookup failed for relation %u", relId);
+	pg_class_tuple = (Form_pg_class) GETSTRUCT(class_tuple);
+
+	dirTablePath = getDirectoryTablePath(spcId, MyDatabaseId, pg_class_tuple->relfilenode);
+
+	ReleaseSysCache(class_tuple);
+
+	/*
+	 * Acquire DirectoryTableLock to ensure that no DROP DIRECTORY TABLE
+	 * or CREATE DIRECTORY TABLE is running concurrently.
+	 */
+	LWLockAcquire(DirectoryTableLock, LW_EXCLUSIVE);
+	if (mkdir(dirTablePath, S_IRWXU) < 0)
+	{
+		ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("unable to create directory \"%s\"", dirTablePath)));
+	}
+	/*
+	 * Advance command counter to ensure the pg_attribute tuple is visible;
+	 * the tuple might be updated to add constraints in previous step.
+	 */
+	CommandCounterIncrement();
+
+	dirRelation = table_open(DirectoryTableRelationId, RowExclusiveLock);
+	/*
+	 * Insert tuple into pg_directory_table.
+	 */
+	memset(values, 0, sizeof(values));
+	memset(nulls, false, sizeof(nulls));
+
+	values[Anum_pg_directory_table_dttablespace - 1] = spcId;
+	values[Anum_pg_directory_table_dtrelid - 1] = ObjectIdGetDatum(relId);
+	values[Anum_pg_directory_table_dtlocation - 1] = CStringGetTextDatum(dirTablePath);
+
+	tuple = heap_form_tuple(dirRelation->rd_att, values, nulls);
+
+	CatalogTupleInsert(dirRelation, tuple);
+
+	heap_freetuple(tuple);
+
+	table_close(dirRelation, RowExclusiveLock);
+
+	LWLockRelease(DirectoryTableLock);
+}
+
+static Datum
+getFileContent(Oid spcId, char *scopedFileUrl)
+{
+	char 	errorMessage[256];
+	char 	buffer[4096];
+	char 	*data;
+	int 	curPos = 0;
+	int		bytesRead;
+	bytea 	*content;
+	UFile	*file;
+	int64 	fileSize;
+
+	file = UFileOpen(spcId,
+				  	 scopedFileUrl,
+				  	 O_RDONLY,
+				  	 errorMessage,
+				  	 sizeof(errorMessage));
+
+	if (file == NULL)
+		ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+					 errmsg("failed to open file \"%s\": %s,", scopedFileUrl, errorMessage)));
+
+	fileSize = UFileSize(file);
+	content = (bytea *) palloc(fileSize + VARHDRSZ);
```

Comment:
alloc file size memory that's really heavy cost of memory.

Reply:
Yes, as now we will form the tuple with file content, it can not avoid. What's more, as the file's size is usually MB magnitude, and we will free the memory after read one file content, it's ok here.

Reply:
well, could you please add a check `if (fileSize > MaxAllocSize)` then goto error.

Reply:
OK